flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,381 @@
+		ResourceManagerMetricGroup resourceManagerMetricGroup) {
+		super(
+			rpcService,
+			resourceManagerEndpointId,
+			resourceId,
+			highAvailabilityServices,
+			heartbeatServices,
+			slotManager,
+			jobLeaderIdService,
+			clusterInformation,
+			fatalErrorHandler,
+			resourceManagerMetricGroup);
+
+		this.flinkConfig = flinkConfig;
+		this.clusterId = flinkConfig.getString(KubernetesConfigOptions.CLUSTER_ID);
+
+		this.workerNodeMap = new ConcurrentHashMap<>();
+
+		this.numPendingPodRequests = 0;
+
+		this.numberOfTaskSlots = flinkConfig.getInteger(TaskManagerOptions.NUM_TASK_SLOTS);
+		this.taskExecutorResourceSpec = TaskExecutorResourceUtils.resourceSpecFromConfig(flinkConfig);
+		this.defaultTaskManagerMemoryMB = taskExecutorResourceSpec.getTotalProcessMemorySize().getMebiBytes();
+		this.defaultCpus = flinkConfig.getDouble(KubernetesConfigOptions.TASK_MANAGER_CPU, numberOfTaskSlots);
+
+		this.slotsPerWorker = createWorkerSlotProfiles(flinkConfig);
+
+		// Load the flink config uploaded by flink client
+		this.flinkClientConfig = GlobalConfiguration.loadConfiguration();
+	}
+
+	@Override
+	protected void initialize() throws ResourceManagerException {
+		LOG.info("Initializing Kubernetes client.");
+		LOG.info("KubernetesResourceManager.initialize clusterId:{}", clusterId);
+
+		this.kubeClient = createFlinkKubeClient();
+
+		try {
+			getWorkerNodesFromPreviousAttempts();
+		} catch (Exception e) {
+			throw new ResourceManagerException(e);
+		}
+
+		this.kubeClient.watchPodsAndDoCallback(getTaskManagerLabels(), this);
+	}
+
+	@Override
+	public CompletableFuture<Void> onStop() {
+		// shut down all components
+		Throwable firstException = null;
+
+		if (kubeClient != null) {
+			try {
+				kubeClient.close();
+			} catch (Throwable t) {
+				firstException = t;
+			}
+		}
+
+		final CompletableFuture<Void> terminationFuture = super.onStop();
+
+		if (firstException != null) {
+			return FutureUtils.completedExceptionally(new FlinkException(
+				"Error while shutting down Kubernetes resource manager", firstException));
+		} else {
+			return terminationFuture;
+		}
+	}
+
+	@Override
+	protected void internalDeregisterApplication(ApplicationStatus finalStatus, @Nullable String optionalDiagnostics) {
+		LOG.info("Stopping kubernetes cluster, id: {}", clusterId);
+		this.kubeClient.stopAndCleanupCluster(this.clusterId);
+	}
+
+	@Override
+	public Collection<ResourceProfile> startNewWorker(ResourceProfile resourceProfile) {
+		LOG.info("Starting new worker with resource profile, {}", resourceProfile.toString());
+		if (!slotsPerWorker.iterator().next().isMatching(resourceProfile)) {
+			return Collections.emptyList();
+		}
+		requestKubernetesPod();
+		return slotsPerWorker;
+	}
+
+	@Override
+	protected KubernetesWorkerNode workerStarted(ResourceID resourceID) {
+		return workerNodeMap.get(resourceID);
+	}
+
+	@Override
+	public boolean stopWorker(final KubernetesWorkerNode worker) {
+		Preconditions.checkNotNull(this.kubeClient);
```

Comment:
No need to check if make `kubeClient` as final.

Reply:
Will remove.